src/libraries/Microsoft.PowerFx.Connectors/OpenApiExtensions.cs

```diff
@@ -0,0 +1,230 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using Microsoft.OpenApi.Any;
+using Microsoft.OpenApi.Models;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Connectors
+{
+    // See definitions for x-ms extensions:
+    // https://docs.microsoft.com/en-us/connectors/custom-connectors/openapi-extensions#x-ms-visibility
+    internal static class OpenApiExtensions
+    {
+        public static string GetBasePath(this OpenApiDocument openApiDocument)
+        {
+            string basePath = null;
+            if (openApiDocument?.Servers.Count == 1)
+            {
+                // This is a full URL that will pull in 'basePath' property from connectors. 
+                // Extract BasePath back out from this. 
+                var fullPath = openApiDocument.Servers[0].Url;
+                var uri = new Uri(fullPath);
+                basePath = uri.PathAndQuery;
+            }
+
+            return basePath;
+        }
+
+        // Get suggested options values.  Returns null if none. 
+        public static string[] GetOptions(this OpenApiParameter param)
+        {
+            // x-ms-enum-values is: array of { value :string, displayName:string}.
+            if (param.Extensions.TryGetValue("x-ms-enum-values", out var value))
+            { 
+                if (value is OpenApiArray array)
+                {
+                    var list = new List<string>(array.Capacity);
+
+                    foreach (var item in array)
+                    {
+                        if (item is OpenApiObject obj)
+                        {
+                            // has keys, "value", and "displayName"
+                            if (obj.TryGetValue("value", out var value2))
+                            {
+                                if (value2 is OpenApiString str)
+                                {
+                                    list.Add(str.Value);
+                                    continue;
+                                }
+                            }
+                        }
+
+                        throw new NotImplementedException($"Unrecognized x-ms-enum-values schema");
+                    }
+
+                    return list.ToArray();
+                }
+            }
+
+            return null;
+        }
+
+        public static bool IsTrigger(this OpenApiOperation op)
+        {
+            var isTrigger = op.Extensions.ContainsKey("x-ms-trigger");
+            return isTrigger;
+        }
+
+        public static bool TryGetDefaultValue(this OpenApiParameter param, out string defaultValue)
```

Comment:
>string [](http://example.com/codeflow?start=79&length=6)

is 'string' the right thing here? That assumes a format. 
 Or should it be FormulaValue / Object?